makefiles/multislot.mk

```diff
@@ -47,8 +47,97 @@ flash-mcuboot: HEXFILE = $(SIGN_BINFILE)
 flash-mcuboot: mcuboot $(FLASHDEPS) flash-bootloader
 	FLASH_ADDR=$(SLOT0_SIZE) $(FLASHER) $(FFLAGS)
 
+###############################################################################
+
+ELFSLOT0        := $(ELFFILE:%.elf=%.slot0.elf)
+ELFSLOT1        := $(ELFFILE:%.elf=%.slot1.elf)
+ELFSLOT2        := $(ELFFILE:%.elf=%.slot2.elf)
+BINSLOT0        := $(ELFSLOT0:%.elf=%.bin)
+BINSLOT1        := $(ELFSLOT1:%.elf=%.bin)
+BINSLOT2        := $(ELFSLOT2:%.elf=%.bin)
+IMAGE_SLOT1     := $(ELFFILE:%.elf=%-slot1-$(APP_ID)-$(APP_VERSION).img)
+IMAGE_SLOT2     := $(ELFFILE:%.elf=%-slot2-$(APP_ID)-$(APP_VERSION).img)
+COMBINED_BIN    := $(ELFFILE:%.elf=%.combined.bin)
+BOOTLOADER_BIN  := $(RIOTBASE)/bootloader/bin/$(BOARD)/bootloader.slot0.bin
+
+FIRMWARE_METADATA_SIZE ?= 0x100
+FIRMWARE_TOOLS = $(RIOTBASE)/dist/tools/firmware
+FIRMWARE       = $(FIRMWARE_TOOLS)/bin/firmware
+GENMETA        = $(FIRMWARE) genmeta
+GENKEYS        = $(FIRMWARE) genkeys
+VERIFY         = $(FIRMWARE) verify
+
+SECKEY ?= $(RIOTBASE)/key.sec
+PUBKEY ?= $(RIOTBASE)/key.pub
+
+$(BOOTLOADER_BIN):
+	@env -i PATH=$(PATH) BOARD=$(BOARD) make -C $(RIOTBASE)/bootloader bootloader
+
+$(FIRMWARE):
+	$(Q)env -i CFLAGS+=-DFIRMWARE_METADATA_SIZE=$(FIRMWARE_METADATA_SIZE) \
+	make clean all -C $(FIRMWARE_TOOLS)
+
+$(SECKEY) $(PUBKEY): $(FIRMWARE)
+	$(Q)$(GENKEYS) $(SECKEY) $(PUBKEY)
+
+bootloader: link
+	@$(_LINK) \
+		$(LINKFLAGPREFIX)--defsym=length="$(SLOT0_SIZE)" \
+		-o $(ELFSLOT0) && \
+		$(OBJCOPY) -Obinary $(ELFSLOT0) $(BINSLOT0) && \
+		truncate -s $$(($(SLOT0_SIZE))) $(BINSLOT0)
+
+multislot: $(SECKEY) $(PUBKEY) link $(BOOTLOADER_BIN)
+	$(Q)$(_LINK) \
+		$(LINKFLAGPREFIX)--defsym=offset="$(SLOT0_SIZE)+$(FIRMWARE_METADATA_SIZE)" \
+		$(LINKFLAGPREFIX)--defsym=length="$(SLOT1_SIZE)-$(FIRMWARE_METADATA_SIZE)" \
+		-o $(ELFSLOT1) && \
+	$(OBJCOPY) -Obinary $(ELFSLOT1) $(BINSLOT1) && \
+	$(GENMETA) $(BINSLOT1) $(APP_VERSION) $(APP_ID) $$(($(SLOT0_ADDR)+$(SLOT0_SIZE)+$(FIRMWARE_METADATA_SIZE))) $(SECKEY) $(BINSLOT1).meta && \
+	cat $(BINSLOT1).meta $(BINSLOT1) > $(IMAGE_SLOT1) && \
+	\
+	$(_LINK) \
+		$(LINKFLAGPREFIX)--defsym=offset="$(SLOT0_SIZE)+$(SLOT1_SIZE)+$$(($(FIRMWARE_METADATA_SIZE)))" \
+		$(LINKFLAGPREFIX)--defsym=length="$(SLOT2_SIZE)-$(FIRMWARE_METADATA_SIZE)" \
+		-o $(ELFSLOT2) && \
+	$(OBJCOPY) -Obinary $(ELFSLOT2) $(BINSLOT2) && \
+	$(GENMETA) $(BINSLOT2) $(APP_VERSION) $(APP_ID) $$(($(SLOT0_ADDR)+$(SLOT0_SIZE)+$(SLOT1_SIZE)+$(FIRMWARE_METADATA_SIZE))) $(SECKEY) $(BINSLOT2).meta && \
+	cat $(BINSLOT2).meta $(BINSLOT2) > $(IMAGE_SLOT2) \
+	; \
+
+combined: multislot
+	$(Q)sh -c 'cat $(BOOTLOADER_BIN) $(IMAGE_SLOT1) > $(COMBINED_BIN)'
+
+.PHONY: flash-multislot verify firmware-tools-clean
+
+ifneq (, $(filter $(BOARD),iotlab-m3))
```

Comment:
Yes I know, it was an initial way to avoid it. We maybe need to decide which kind of files we prefer to flash in a board. We can start to put binary by default in boards on which the bootloader can be used.